Devise を使用したユーザー認証機能とユーザーページを作成する

nakamu-kazu222 opened the pull request:
- Devise gemを使ってユーザー認証機能を実装する
- 各ユーザーのプロフィールページを作る
    - 郵便番号、住所と自己紹介文を表示できるようにする

nakamu-kazu222 committed:
Devise のインストール

nakamu-kazu222 committed:
Deviseの初期設定

rails g devise:install実行で関連ファイルを作成した

nakamu-kazu222 committed:
DeviseのMailerで使用するデフォルトURLのみ設定した

nakamu-kazu222 committed:
Devise で扱う User モデルを作成した

nakamu-kazu222 committed:
マイグレーションを実行した

nakamu-kazu222 committed:
ヘルパーメソッドを使ってユーザー認証を実装した

nakamu-kazu222 committed:
Devise の View を作成した

nakamu-kazu222 committed:
Devise の UserController を作成した

nakamu-kazu222 committed:
user テーブルにカラムを追加した

Deviseの初期状態でのカラムに「name」「profile」カラムがないため追加した

nakamu-kazu222 committed:
バリデーションを追加した

nakamu-kazu222 committed:
ログイン済ユーザーのみにアクセスを許可する

nakamu-kazu222 committed:
View を編集した

ユーザー情報を記入
ユーザー情報を編集

nakamu-kazu222 committed:
コントローラーとviewファイルを生成した

Deviseとまた別にコントローラーとshow.html.erbを作成した

nakamu-kazu222 committed:
Modelごとのviewファイルを有効化した

nakamu-kazu222 committed:
ログイン、サインイン機能を追加した

ユーザーがログインしないとアプリを使用できないようにした
ユーザー情報を表示できるようにした
ログアウトできるようにした

nakamu-kazu222 committed:
ユーザー一覧画面を表示できるようにした

nakamu-kazu222 committed:
ユーザー情報を編集するようにした

nakamu-kazu222 committed:
ページング処理を追加した

nakamu-kazu222 committed:
i18nで日本語化した

Deviseが用意した画面を日本語で表示した
新しく追加した画面を日本語で表示した

nakamu-kazu222 committed:
パスワードを忘れたらパスワード再設定メールを経由してパスワードを再設定できるようにした

送信したメールをブラウザ上で確認できる

nakamu-kazu222 committed:
郵便番号、住所、自己紹介を保存できるようにした

nakamu-kazu222 committed:
Rubocopによる修正をした

nakamu-kazu222 committed:
ERB Lint実行で修正をした

yuuu commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,17 @@
 # frozen_string_literal: true
 
 class ApplicationController < ActionController::Base
+  protect_from_forgery with: :exception
```
こちらを追加した理由を教えていただけますでしょうか？

nakamu-kazu222 replied:
こちらの[サイト](https://qiita.com/k19911848/items/fe5f48b3d2c902ac07a2)を参考にして実装した際に、CSRF対策をした方がいいと思い追加しました

yuuu replied:
なるほど。参照しているドキュメントが古いですね。Rails 5.2以降このコードは不要となっていますので削除をお願いします 🙏 
参考: https://railsguides.jp/upgrading_ruby_on_rails.html#protect-from-forgery%E3%81%AF%E4%BB%8A%E5%BE%8C%E3%83%87%E3%83%95%E3%82%A9%E3%83%AB%E3%83%88%E3%81%A7prepend-false%E3%81%AB%E8%A8%AD%E5%AE%9A%E3%81%95%E3%82%8C%E3%82%8B

Qiitaなどを参考にすること自体は悪くないですが、公式ドキュメント(今回だとDeviseのREADME)と照らし合わせながら手順を確認することをおすすめします 👍 公式ドキュメントにはこの手順が存在しないので、不要な手順であると気づくことができるかと。

nakamu-kazu222 replied:
Rails のバージョンは気づきませんでした
公式ドキュメント(Railsガイドなど)を確認しつつ、他のサイトを参考にするようにします

yuuu commented on app/controllers/users/application_controller.rb:
```diff
@@ -0,0 +1,7 @@
+
+class ApplicationController < ActionController::Base
+  def after_sign_in_path_for
+    blogs_path
```
このシステムに `blogs_path` って存在していますか？

nakamu-kazu222 replied:
存在していません
実装に関係ないので削除します

yuuu commented on app/controllers/users/confirmations_controller.rb:
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Users::ConfirmationsController < Devise::ConfirmationsController
```
特にカスタマイズがないのであれば削除しましょう。これらのファイルは削除しても動作は変わらないはずです。

nakamu-kazu222 replied:
こちら不要ですね
削除します

yuuu commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  before_action :set_user, only: %i[show edit update]
+
+  def index
+    @users = User.page(params[:page]).per(2)
```
明示的に `order` がないので表示順がリロードのたびに変わる懸念があります。追加をお願い致します。

nakamu-kazu222 replied:
orderメソッドで、ユーザーを作成日時の昇順に並べ替えるように変更しました

yuuu commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,38 @@
+  end
+
+  def update
+    @user = current_user
```
この行は必要でしょうか？むしろ `set_user` でセットされた `@user` が `current_user` かどうかをチェックすべきではないでしょうか？

nakamu-kazu222 replied:
`set_user`に変更して、`@user`が `current_user`かをチェックするようにしました

yuuu commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,38 @@
+    @user = current_user
+
+    if @user.update(user_params)
+      redirect_to root_path, notice: 'Your account information has been updated'
```
i18n対応が漏れているようなので対応をお願い致します

nakamu-kazu222 replied:
翻訳ファイルとともに対応しました

yuuu commented on app/helpers/users_helper.rb:
```diff
@@ -0,0 +1,4 @@
+# frozen_string_literal: true
+
+module UsersHelper
```
こちらも不要であれば削除しておきましょう

nakamu-kazu222 replied:
不要なため削除しました

yuuu commented on config/application.rb:
```diff
@@ -11,6 +11,8 @@ class Application < Rails::Application
     # Initialize configuration defaults for originally generated Rails version.
     config.load_defaults 7.0
 
+    config.i18n.load_path += Dir[Rails.root.join('config', 'locales', '**', '*.{rb,yml}').to_s]
```
この記述は必要でしょうか？追加した意図があれば教えてください 🙏

nakamu-kazu222 replied:
将来的に他の言語を翻訳することになった場合を想定して記述しました
しかし、こちらの行がなくても`config/locales/translation_ja.yml`を読み込むので、削除します

yuuu replied:
削除すること自体は問題ないのですが、この行で実現できることを勘違いしてそうですね。
`locales` ディレクトリ以下にサブディレクトリを作成して、YAMLファイルが階層管理される場合はこのコードが必要となります。他の言語に対応するだけであればこのコードは不要です。

参考: https://railsguides.jp/i18n.html#%E3%83%AD%E3%82%B1%E3%83%BC%E3%83%AB%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB%E3%81%AE%E7%B7%A8%E6%88%90

nakamu-kazu222 replied:
> locales ディレクトリ以下にサブディレクトリを作成して、YAMLファイルが階層管理される場合はこのコードが必要

勘違いしていました
階層管理されている場合は必要で、他の言語に対応するだけであれば不要ですね

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+  # Configure the e-mail address which will be shown in Devise::Mailer,
+  # note that it will be overwritten if you use your own mailer class
+  # with default "from" parameter.
+  config.mailer_sender = 'please-change-me-at-config-initializers-devise@example.com'
```
`please-change-me` とあるので変更をお願い致します。 `no-reply` のようにすることが多いですね。

nakamu-kazu222 replied:
Devise から送信されるメールの送信元メールアドレスを変更しました

yuuu commented on config/initializers/devise.rb:
```diff
@@ -0,0 +1,313 @@
+
+  # ==> Configuration for :validatable
+  # Range for password length.
+  config.password_length = 6..128
```
昨今のセキュリティ情勢を考えると6文字のパスワードは文字数が少ないですね。

nakamu-kazu222 replied:
10文字に変更しました

yuuu commented on config/routes.rb:
```diff
@@ -1,7 +1,23 @@
+  end
+
+  devise_for :users, controllers: { registrations: 'users/registrations' }
+  get "users/index" => "users#index", as: :users
```
ここは `resources` を使うようにしましょう。

nakamu-kazu222 replied:
`resource`でルーティング設定をするように変更しました

yuuu commented on test/fixtures/users.yml:
```diff
@@ -0,0 +1,11 @@
+# Read about fixtures at https://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
このあたりのテストファイルも、使う予定がないのであれば削除しておいてください。

nakamu-kazu222 replied:
使う予定がないので削除しました

nakamu-kazu222 committed:
記述しなくてもtranslation_ja.ymlを読み込むので、削除した

nakamu-kazu222 committed:
認証キーでユーザー名は指定しなくても、メールアドレスが一意な値であるため変更した

nakamu-kazu222 committed:
実装に関係ないので削除した

nakamu-kazu222 committed:
不要なファイルなため削除した

nakamu-kazu222 committed:
表示順がリロードのたびに変わる懸念があるため、ユーザーを作成日時の昇順に並べ替えるように変更した

nakamu-kazu222 committed:
userが current_userかをチェックするようにした

nakamu-kazu222 committed:
i18n対応をした

nakamu-kazu222 committed:
Devise から送信されるメールの送信元メールアドレスを変更した

nakamu-kazu222 committed:
昨今のセキュリティ情勢を考え、パスワードの文字数を10文字以上に変更した

nakamu-kazu222 committed:
resourceを使用して、ルーティング設定した

yuuu commented on app/controllers/users/omniauth_callbacks_controller.rb:
```diff
@@ -0,0 +1,30 @@
+# frozen_string_literal: true
+
+class Users::OmniauthCallbacksController < Devise::OmniauthCallbacksController
```
これらのコントローラーについても特にカスタマイズがないのであれば、不要なファイルなはずなので削除をお願いします。

nakamu-kazu222 replied:
カスタマイズしていないので削除しました

yuuu commented on app/controllers/users/registrations_controller.rb:
```diff
@@ -0,0 +1,66 @@
+  # end
+
+  def after_update_path_for
+    user_path(id: current_user.id)
```
ここは `user_path(current_user)` と書けますね。

nakamu-kazu222 replied:
`user_path(current_user)`に変更しました

yuuu commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,28 @@
+  # end
+
+  # POST /resource/sign_in
+  def create; end
```
ここは `super` しなくて良いですか？

nakamu-kazu222 replied:
super を呼び出して、親クラスである Devise::SessionsController の create アクションを実行するようにしました

yuuu commented on app/views/devise/mailer/confirmation_instructions.html.erb:
```diff
@@ -0,0 +1,5 @@
+<p><%= t('.greeting', recipient: @email) %></p>
```
今回 `confirmation` は行わない認識ですので、 `confirmation` 関連のファイルは削除しておいてください。 `unlock` についても同様です。

nakamu-kazu222 replied:
confirmation 関連のファイルと unlock関連のファイルを削除しました

nakamu-kazu222 committed:
Rails 5.2以降このコードは不要となっているので削除した

nakamu-kazu222 committed:
カスタマイズしなく、不要なファイルなので削除した

nakamu-kazu222 committed:
password_confirmationは不要なので削除した

nakamu-kazu222 committed:
user_path(current_user)を記載できる

nakamu-kazu222 committed:
super を呼び出した
親クラスである Devise::SessionsController の create アクションを実行するように変更した

nakamu-kazu222 committed:
create アクションがコントローラー内で定義されていない

create アクションが空の状態でオーバーライドされているため、明示的な定義をした

yuuu commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,32 @@
+  # POST /resource/sign_in
+
+  def create
+    super do |resource|
```
ブロックを渡しているのはなぜでしょうか？

nakamu-kazu222 replied:
誤った知識でした
学習し直しましたが、今回は親クラスの create メソッドを呼び出す際に、その前にブロック内の処理はいらないので、ブロックは削除します

yuuu commented on app/views/users/edit.html.erb:
```diff
@@ -0,0 +1,11 @@
+<h1><%= t('views.users.edit.title') %></h1>
```
この翻訳は遅延探索を使うようにしましょう。その方がスッキリ書けるはずです。
https://railsguides.jp/i18n.html#%E9%81%85%E5%BB%B6%E6%8E%A2%E7%B4%A2

yuuu replied:
他のページのタイトルも同様です 🙏

nakamu-kazu222 replied:
遅延探索を使用して翻訳の定義をしました
遅延探索に対応するためYAMLも変更しました

nakamu-kazu222 committed:
今回は親クラスの create メソッドを呼び出す際に、その前にブロック内の処理はないので変更した

nakamu-kazu222 committed:
翻訳に遅延探索を使用した

遅延探索に対応するためYAMLを変更した
各ページに遅延探索を対応した

yuuu commented on app/controllers/users/sessions_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  # end
+
+  # POST /resource/sign_in
+  def create
```
`super` を呼ぶだけであれば、他のメソッド同様にコメントアウトしても動作するのではないでしょうか？ 🤔

nakamu-kazu222 replied:
おっしゃる通りで、コメントアウトしても動作します
コメントアウトして対応しました

また、createメソッドをフィルターかける必要がなくなったため、befor_actionも同様にコメントアウトしました

yuuu commented on app/views/users/show.html.erb:
```diff
@@ -0,0 +1,12 @@
+<h1><%= t('.title') %></h1>
+
+<h3><%= t('.user_name') %></h3>
```
このあたりはi18nのプラクティスでも学ばれたと思いますが、ActiveRecordのカラムの翻訳なので、遅延探索ではなく専用のメソッドを使うようにしましょう。
https://railsguides.jp/i18n.html#active-record%E3%83%A2%E3%83%87%E3%83%AB%E3%81%A7%E7%BF%BB%E8%A8%B3%E3%82%92%E8%A1%8C%E3%81%AA%E3%81%86

遅延探索はやみくもに使うものではなく、「そのページでしか使わないであろう翻訳」に使うようにしてください。そうしないと、同じ翻訳の定義を何度もYAMLファイル中に記述することになってしまいます。

nakamu-kazu222 replied:
> そのページでしか使わないであろう翻訳」に使う

遅延探索の使用する箇所を理解していませんでした
今回は、タイトルに関しての翻訳で遅延探索を使用するように変更しました

また、複数のページで翻訳する内容は、"Model.human_attribute_name(attribute)メソッド"を使用して翻訳するように変更しました

nakamu-kazu222 committed:
コメントアウトしても動作するので対応した

nakamu-kazu222 committed:
createメソッドは、super を呼ぶだけであれば動作するのでコメントアウトした

createメソッドをコメントアウトしたので、before_actionも同様にコメントアウトした

nakamu-kazu222 committed:
翻訳情報を変更した

遅延探索は、そのページでしか使わない翻訳に使用した
それ以外の複数のページで使用する翻訳は、Model.human_attribute_name(attribute)メソッドを使用した

nakamu-kazu222 committed:
ユーザーコントローラーをリファクタリングした

set_user はbefor_actionで指定しているので、updateメソッドではいらない
 また、showメソッドもbefor_actionで呼んでいるのでいらない
